additional testing of DKG flow

CluEleSsUK opened the pull request:
- fixed some PR comments from a previous PR

CluEleSsUK committed:
additional testing of DKG flow

- fixed some PR comments from a previous PR

AnomalRoil commented on sidecar/internal/cmd/start.go:
```diff
@@ -65,7 +65,7 @@ func Start(_ *cobra.Command, _ []string) {
-		slog.Error("error running daemon", err)
+		slog.Error("error running daemon", "err", err)
 		os.Exit(1)
 	}
```
This will os.Exit on the first read on the channel, so why the for loop?
Even in case of a `nil`

AnomalRoil commented on integration/integration_test.go:
```diff
@@ -42,35 +81,78 @@ func startStubSSVNode(t *testing.T, ssvPort uint) {
 	err := sidecar.GenerateKey(keyPath)
 	if err != nil {
-		return sidecar.Daemon{}, err
+		t.FailNow()
```
```suggestion
		t.Fatal(err)
```

AnomalRoil commented on integration/integration_test.go:
```diff
@@ -42,35 +81,78 @@ func startStubSSVNode(t *testing.T, ssvPort uint) {
 	_, err = sidecar.SignKey(url, keyPath)
 	if err != nil {
-		return sidecar.Daemon{}, err
+		t.FailNow()
```
```suggestion
		t.Fatal(err)
```

AnomalRoil commented on integration/integration_test.go:
```diff
@@ -42,35 +81,78 @@ func startStubSSVNode(t *testing.T, ssvPort uint) {
-	return sidecar.NewDaemon(port, url, ssvURL, keyPath)
+	d, err := sidecar.NewDaemon(port, url, ssvURL, keyPath)
+	if err != nil {
+		t.FailNow()
```
> Fatal is equivalent to `Log` followed by `FailNow`. 

```suggestion
		t.Fatal(err)
```

AnomalRoil commented on shared/crypto/dkg.go:
```diff
@@ -26,6 +26,21 @@ func MarshalDistKey(key *share.PriShare) ([]byte, error) {
+
+	V := scheme.KeyGroup().Scalar()
+	err := V.UnmarshalBinary(bytes[8:])
+	if err != nil {
```
```suggestion
	if err := V.UnmarshalBinary(bytes[8:]); err != nil {
```

AnomalRoil commented on sidecar/daemon.go:
```diff
@@ -54,13 +60,12 @@ func NewDaemon(port uint, publicURL string, ssvURL string, keyPath string) (Daem
 	slog.Info(fmt.Sprintf("Keypair loaded from %s", keyPath))
 
 	thresholdScheme := crypto.NewBLSSuite()
-	dkgCoordinator := dkg.NewDKGCoordinator(publicURL, thresholdScheme)
 	daemon := Daemon{
 		port:             port,
 		key:              keypair,
 		publicURL:        publicURL,
 		ssvClient:        api.NewSsvClient(ssvURL),
-		dkg:              &dkgCoordinator,
+		dkg:              coordinator,
```
I'm not a huge fan of really passing everything to be copied around like that, especially fat structs, but I guess it's fine

AnomalRoil replied:
Especially since your functions on the Coordinator are taking pointer receivers...

AnomalRoil commented on sidecar/dkg/dkg.go:
```diff
@@ -15,21 +15,26 @@ import (
 
-func NewDKGCoordinator(publicURL string, scheme crypto.ThresholdScheme) DKGCoordinator {
-	return DKGCoordinator{
+func NewDKGCoordinator(publicURL string, scheme crypto.ThresholdScheme) Coordinator {
+	return Coordinator{
```
Shouldn't this return a pointer given the rest of the functions here are taking point receivers?

CluEleSsUK committed:
changed tests to Fatal()

pointer on creation of dkg coordinator

CluEleSsUK committed:
obeying daddy linter

CluEleSsUK committed:
some more fixes from other PR review

AnomalRoil commented on shared/crypto/dkg.go:
```diff
@@ -26,6 +26,20 @@ func MarshalDistKey(key *share.PriShare) ([]byte, error) {
 }
 
+func UnmarshalDistKey(scheme ThresholdScheme, bytes []byte) (share.PriShare, error) {
+	I := binary.BigEndian.Uint64(bytes[0:8])
```
This could cause a remote crash I think:
https://go.dev/play/p/PSO1zKw66p9

> panic: runtime error: slice bounds out of range [:8] with capacity 5

You might want to check the length first.

CluEleSsUK replied:
updated

CluEleSsUK replied:
added a test too for posterity

AnomalRoil commented on sidecar/dkg/dkg.go:
```diff
@@ -15,21 +15,26 @@ import (
 )
 
-type DKGCoordinator struct {
+type Protocol interface {
+	RunDKG(identities []crypto.Identity, sessionID []byte, keypair crypto.Keypair) (*dkg.Result, error)
+	ProcessPacket(packet api.SidecarDKGPacket) error
+}
```
In Go interfaces usually are best defined on the receiver side.

https://medium.com/@cep21/preemptive-interface-anti-pattern-in-go-54c18ac0668a

CluEleSsUK replied:
updated

CluEleSsUK committed:
added some length checks to marshalling code

moved dkg.Protocol interface to sidecar

CluEleSsUK committed:
added a test for the distkey marshalling

CluEleSsUK committed:
added some test cleanup so ports aren't held

AnomalRoil commented on shared/crypto/dkg.go:
```diff
@@ -58,6 +76,11 @@ func UnmarshalPubPoly(scheme ThresholdScheme, b []byte) (share.PubPoly, error) {
 	pointLen := group.PointLen()
 	base := group.Point()
+
+	if len(b) < pointLen {
```
I feel like this might not protect from crashes the below 
```
for i := pointLen; i < len(b); i += pointLen {
	p := group.Point()
	err = p.UnmarshalBinary(b[i : i+pointLen])
```

What if `len(b) == 20` and `pointLen == 12` for instance? 

Since the for loop is working as follows:
 -   the init statement: executed before the first iteration
 -   the condition expression: evaluated before every iteration
 -   the post statement: executed at the end of every iteration

My understanding is that we'd have:
1. `i := 12` ✅ 
2. `i < 20`  ✅ 
3. `b[12 : 24]` --> 24 is out of bound => crash ❌

CluEleSsUK replied:
fixed, good catch!

AnomalRoil commented on sidecar/daemon.go:
```diff
@@ -19,13 +20,24 @@ type Daemon struct {
 }
 
+type DKGProtocol interface {
+	RunDKG(identities []crypto.Identity, sessionID []byte, keypair crypto.Keypair) (*kyber.Result, error)
```
I feel like we might want to stay away from concrete Kyber types in this repo, so maybe use an interface instead of `kyber.Result` in order to allow us to plug another DKG easily in the future if needed?

CluEleSsUK replied:
I've taken the kyber types out of the sidecar now. The downside is we end up juggling back and forth between `[]byte` when we use the scheme, but it's not a big deal

AnomalRoil commented on sidecar/internal/cmd/start.go:
```diff
@@ -63,9 +63,7 @@ func Start(_ *cobra.Command, _ []string) {
-		slog.Error("error running daemon", err)
-		os.Exit(1)
-	}
+	err = <-errs
+	slog.Error("error running daemon", "err", err)
+	os.Exit(1)
```
So, no graceful exit option in case of no errors?

CluEleSsUK replied:
Have now checked for `ErrServerClosed`

CluEleSsUK committed:
bounds check when unmarshalling polynomials

CluEleSsUK committed:
pr comments

- removed specific kyber structs from some interfaces
- graceful shutdown if Stop() is called
